src/misc/npkit.cc

```diff
@@ -17,19 +17,30 @@ NpKitEvent** NpKit::cpu_event_buffers_ = nullptr;
 
 NpKitEventCollectContext* NpKit::gpu_collect_contexts_ = nullptr;
 NpKitEventCollectContext* NpKit::cpu_collect_contexts_ = nullptr;
-uint64_t* NpKit::cpu_timestamp_ = nullptr;
 
-std::thread* NpKit::cpu_timestamp_update_thread_ = nullptr;
-volatile bool NpKit::cpu_timestamp_update_thread_should_stop_ = false;
+uint64_t NpKit::base_cpu_timestamp_global_ = 0;
+uint64_t NpKit::base_cpu_timestamp_local_ = 0;
+uint64_t NpKit::base_gpu_timestamp_cpu_ = 0;
+uint64_t NpKit::base_gpu_timestamp_gpu_ = 0;
+
+__global__ void TimeCalibrationKernel(uint64_t *gpu_timestamp, uint64_t *cpu_timestamp_in, uint64_t *cpu_timestamp_out) {
+  uint64_t gpu_timestamp_1 = NPKIT_GET_GPU_TIMESTAMP();
+  uint64_t cpu_timestamp = *cpu_timestamp_in;
+  uint64_t gpu_timestamp_2 = NPKIT_GET_GPU_TIMESTAMP();
+  *gpu_timestamp = gpu_timestamp_1 + (gpu_timestamp_2 - gpu_timestamp_1) / 2;
+  *cpu_timestamp_out = cpu_timestamp;
+}
+
+static volatile bool cpu_timestamp_update_thread_should_stop = false;
+static volatile uint64_t* volatile_cpu_timestamp = nullptr;
 
 void NpKit::CpuTimestampUpdateThread() {
```

Comment:
In general, since RTC is available on GFX9 and GFX10, we would prefer to use known RTC clock instead of CPU time stamp. This is more accurate and also save system resources (no extra CPU thread).

Reply:
Currently there is no CPU timestamp thread anymore. This is only used in initialization phase.